ModelViewer/ModelViewer/ModelLoader.cpp

```diff
@@ -25,154 +26,178 @@ Model ModelLoader::LoadModel(const QString& file)
+	AssignColor(newMesh.m_ambient, color);
+	pMaterial->Get(AI_MATKEY_COLOR_DIFFUSE, color);
+	AssignColor(newMesh.m_diffuse, color);
+	pMaterial->Get(AI_MATKEY_COLOR_SPECULAR, color);
+	AssignColor(newMesh.m_specular, color);
+	pMaterial->Get(AI_MATKEY_SHININESS, val);
+	newMesh.m_shininess = val;
+
+
+
+	// TODO: Handle textures
+
+
+
+	// Determine if we have colors, normals, or texture coordinates
+	newMesh.m_hasNormals = (pMesh->mNormals != NULL);
+	newMesh.m_hasUVCoordinates = pMesh->HasTextureCoords(0);
+	newMesh.m_hasColors = pMesh->HasVertexColors(0);
+
+	// Determine the size of a single attribute
+	newMesh.m_numPositionComponents = 3;
+	newMesh.m_numNormalComponents = 3;
+	newMesh.m_numUVComponents = pMesh->mNumUVComponents[0];  // Various
+	newMesh.m_numColorComponents = 4;
+
+	// Compute sizes for all attributes
+	const int positionBufferSize = newMesh.m_numPositionComponents * pMesh->mNumVertices * sizeof(float);
+	const int normalBufferSize = newMesh.m_numNormalComponents * pMesh->mNumVertices * sizeof(float);
+	const int uvBufferSize = newMesh.m_numUVComponents * pMesh->mNumVertices * sizeof(float);
+	const int colorBufferSize = newMesh.m_numColorComponents * pMesh->mNumVertices * sizeof(float);
+
+	// Compute the offsets of features in the overall vertex buffer
+	//   1. Positions (required)
+	//   2. Normals (optional)
+	//   3. UV Coordinates (optional)
+	//   4. Colors (optional)
+	int currentOffset = 0;
+	if (true) {
+		newMesh.m_positionOffset = currentOffset;
+		currentOffset += positionBufferSize;
+	}
+	if (newMesh.m_hasNormals) {
+		newMesh.m_normalOffset = currentOffset;
+		currentOffset += normalBufferSize;
+	}
+	if (newMesh.m_hasUVCoordinates) {
+		newMesh.m_uvOffset = currentOffset;
+		currentOffset += uvBufferSize;
 	}
+	if (newMesh.m_hasColors) {
+		newMesh.m_colorOffset = currentOffset;
+		currentOffset += colorBufferSize;
+	}
+	const int vertexBufferTotalSize = currentOffset;
+
+	// Allocate the vertex buffer all at once to save time
+	newMesh.m_vertexBuffer.allocate(vertexBufferTotalSize);
+
+	// Positions
+	newMesh.m_vertexBuffer.write(newMesh.m_positionOffset, pMesh->mVertices, positionBufferSize);
+
+	// Normals
+	if (newMesh.m_hasNormals) {
+		newMesh.m_vertexBuffer.write(newMesh.m_normalOffset, pMesh->mNormals, normalBufferSize);
+	}
+
+	// UV Coordinates
+	if (newMesh.m_hasUVCoordinates) {
+		newMesh.m_vertexBuffer.write(newMesh.m_uvOffset, pMesh->mTextureCoords[0], uvBufferSize);
+	}
+
+	// Colors
+	if (newMesh.m_hasColors)
+	{
+		newMesh.m_vertexBuffer.write(newMesh.m_colorOffset, pMesh->mColors[0], colorBufferSize);
+	}
+
+
+	// Load indices
+	newMesh.m_indexCount = pMesh->mNumFaces * 3;
+	const int indexBufferSize = sizeof(int) * newMesh.m_indexCount;
+	newMesh.m_indexBuffer.allocate(indexBufferSize);
+
+	std::vector<int> indices;
+	indices.reserve(newMesh.m_indexCount);
+
+	// Loop through all faces
+	// Potentially use OpenMP to multithread this if it is slow
```

Comment:
Added a note about this to issue #10

Reply:
Thanks :)

Reply:
I tested some load times against 3D Viewer.  We can already load big models in roughly half the time :)